lib/python/ensembl/brc4/runnable/dump_stable_ids.py

```diff
@@ -0,0 +1,420 @@
+                    continue
+                self.add_pair(pair)
+
+
+class dump_stable_ids:
+
+    def __init__(self, server: str) -> None:
+        self.server = server
+    
+    def dump_history_json(self, output_file: str):
+        pass
+
+    def get_history(self) -> Any:
+        
+        sessions = self.get_mapping_sessions()
+
+        events = []
+        for session in sessions:
+            print(f"Mapping session {session['release']}")
+            pairs = self.get_pairs(session["id"])
+            session_events = self.make_events(pairs)
+            for event in session_events:
+                event.set_release(session["release"])
+                event.set_date(session["date"])
+            events += session_events
+        
+        # Then analyse the pairs to make events
+        return events
+    
+    def print_events(self, events: List[StableIdEvent], output_file: str) -> None:
+        if not events:
+            print("No events to print")
+            return
+        with open(output_file, 'w') as out_fh:
+            for event in events:
+                event_lines = event.brc_format_2()
+                for line in event_lines:
+                    out_fh.write(line + "\n")
+    
+    def get_mapping_sessions(self):
+        query = """SELECT mapping_session_id, new_release, created
+        FROM mapping_session
+        """
+        cursor = self.server.db.cursor()
+        cursor.execute(query)
+
+        sessions = []
+        for db in cursor:
+            date = db[2]
+            session = {"id": db[0], "release": db[1], "date": date}
+            sessions.append(session)
+        return sessions
+
+    def get_pairs(self, session_id):
+        query = """SELECT old_stable_id, new_stable_id
+        FROM stable_id_event
+        WHERE (old_stable_id != new_stable_id OR old_stable_id IS NULL OR new_stable_id IS NULL)
+            AND type="gene"
+            AND mapping_session_id=%s
+        """
+        values = (session_id,)
+        cursor = self.server.db.cursor()
+        cursor.execute(query, values)
+
+        pairs = []
+        for db in cursor:
+            pair = {"old_id": db[0], "new_id": db[1]}
+            pairs.append(pair)
+        print(f"{len(pairs)} stable id events")
+        return pairs
+    
+    def make_events(self, pairs):
+
+        from_list = {}
+        to_list = {}
+        for pair in pairs:
+            old_id = pair["old_id"]
+            new_id = pair["new_id"]
+
+            if old_id in from_list:
+                from_list[old_id].append(new_id)
+            else:
+                from_list[old_id] = [new_id]
+
+            if new_id in to_list:
+                to_list[new_id].append(old_id)
+            else:
+                to_list[new_id] = [old_id]
+        
+        # Remove empty elements
+        for from_id in from_list:
+            from_id_list = from_list[from_id]
+            from_id_list = [id for id in from_id_list if id]
```

Comment:
You could do a generic method to remove empty elements from lists if you want to use it everywhere.

Reply:
using a generic clean_list method now